File: src/main/java/routing_game/Controller/BellmanFordSession.java
package routing_game.Controller;

import routing_game.Controller.BoardView.CheckResult;
import routing_game.Controller.BoardView.SelectResult;
import routing_game.Model.NetworkLink;
import routing_game.Model.NetworkMap;
import routing_game.Model.NetworkNode;
import routing_game.Model.RoutingEntry;
import routing_game.Model.ShortestPaths;

import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.Set;

public final class BellmanFordSession implements BoardView {
    public enum Phase {PICK_LINK, REPAIR, DONE}

    public enum FailResult {OK, ALREADY_FAILED, DISCONNECTS, NOT_USED}

    private final NetworkMap original;
    private NetworkMap live;
    private NetworkLink failed;
    private final Map<Integer, Map<Integer, RoutingEntry>> originalTables = new LinkedHashMap<>();
    private final Map<Integer, Map<Integer, RoutingEntry>> accepted = new HashMap<>();
    private final Map<Integer, Set<Integer>> affected = new LinkedHashMap<>();
    private final Set<Integer> completed = new HashSet<>();
    private Integer selectedNode;
    private Phase phase = Phase.PICK_LINK;

    public BellmanFordSession(int nodeCount, long seed) {
        Random random = new Random(seed);
        NetworkMap chosen = null;
        Map<Integer, Map<Integer, RoutingEntry>> tables = null;
        for (int attempt = 0; attempt < 20; attempt++) {
            NetworkMap candidate = NetworkMap.random(nodeCount, random);
            Map<Integer, Map<Integer, RoutingEntry>> built = hostTables(candidate);
            if (hasInterestingFailure(candidate, built)) {
                chosen = candidate;
                tables = built;
                break;
            }
            if (chosen == null) {
                chosen = candidate;
                tables = built;
            }
        }
        this.original = chosen;
        this.live = original;
        originalTables.putAll(tables);
    }

    public Phase phase() {
        return phase;
    }

    public int remainingRepairs() {
        int n = 0;
        for (Integer id : affected.keySet()) {
            if (!completed.contains(id)) {
                n++;
            }
        }
        return n;
    }

    public int affectedRouterCount() {
        return affected.size();
    }

    public FailResult failLink(NetworkLink link) {
        if (failed != null) {
            return FailResult.ALREADY_FAILED;
        }
        NetworkLink canonical = original.linkBetween(link.a(), link.b());
        if (canonical == null) {
            return FailResult.NOT_USED;
        }
        NetworkMap next = original.withoutLink(canonical);
        if (!next.isConnected()) {
            return FailResult.DISCONNECTS;
        }
        Map<Integer, Set<Integer>> broken = new LinkedHashMap<>();
        for (NetworkNode router : original.routers()) {
            Set<Integer> dests = new LinkedHashSet<>();
            for (NetworkNode host : original.hosts()) {
                if (walkUsesLink(original, originalTables, router.id(), host.id(), canonical)) {
                    dests.add(host.id());
                }
            }
            if (!dests.isEmpty()) {
                broken.put(router.id(), dests);
            }
        }
        if (broken.isEmpty()) {
            return FailResult.NOT_USED;
        }
        this.failed = canonical;
        this.live = next;
        this.affected.putAll(broken);
        this.phase = Phase.REPAIR;
        for (NetworkNode router : original.routers()) {
            if (!affected.containsKey(router.id())) {
                completed.add(router.id());
                accepted.put(router.id(), copyTable(originalTables.get(router.id())));
            }
        }
        return FailResult.OK;
    }

    @Override
    public NetworkMap map() {
        return original;
    }

    @Override
    public Integer selectedNode() {
        return selectedNode;
    }

    @Override
    public SelectResult selectNode(int nodeId) {
        if (original.node(nodeId).isHost()) {
            return SelectResult.HOST;
        }
        selectedNode = nodeId;
        return SelectResult.OK;
    }

    @Override
    public CheckResult check(int nodeId, Map<Integer, RoutingEntry> playerTable) {
        if (failed == null) {
            return CheckResult.INCOMPLETE;
        }
        if (original.node(nodeId).isHost()) {
            return CheckResult.INCOMPLETE;
        }
        if (completed.contains(nodeId)) {
            return CheckResult.CORRECT;
        }
        Set<Integer> dests = affected.get(nodeId);
        if (dests == null || dests.isEmpty()) {
            return CheckResult.CORRECT;
        }
        for (Integer dest : dests) {
            if (playerTable.get(dest) == null) {
                return CheckResult.INCOMPLETE;
            }
        }
        for (Integer dest : dests) {
            RoutingEntry player = playerTable.get(dest);
            if (!ShortestPaths.isLegalWeightedNextHop(live, nodeId, player.nextHop(), dest)) {
                return CheckResult.WRONG_HOP;
            }
        }
        Map<Integer, RoutingEntry> table = copyTable(originalTables.get(nodeId));
        for (Integer dest : dests) {
            int hop = playerTable.get(dest).nextHop();
            int cost = ShortestPaths.weightedDistance(live, nodeId, dest);
            table.put(dest, new RoutingEntry(dest, hop, cost));
        }
        accepted.put(nodeId, table);
        completed.add(nodeId);
        if (remainingRepairs() == 0) {
            phase = Phase.DONE;
        }
        return CheckResult.CORRECT;
    }

    @Override
    public boolean isCompleted(int nodeId) {
        return completed.contains(nodeId);
    }

    @Override
    public boolean isUnlocked(int nodeId) {
        return original.node(nodeId).isRouter();
    }

    @Override
    public Integer toldNextHop(int routerId) {
        Map<Integer, RoutingEntry> table = visibleTable(routerId);
        if (table == null) {
            return null;
        }
        RoutingEntry entry = table.get(original.destId());
        return entry == null ? null : entry.nextHop();
    }

    @Override
    public boolean packetArrived() {
        return false;
    }

    @Override
    public boolean playerPathSuboptimal() {
        return false;
    }

    @Override
    public List<Integer> playerPath() {
        return List.of();
    }

    @Override
    public List<Integer> optimalPath() {
        return List.of();
    }

    @Override
    public int packetNode() {
        return original.sourceId();
    }

    @Override
    public Map<Integer, RoutingEntry> acceptedTable(int nodeId) {
        return accepted.get(nodeId);
    }

    @Override
    public Map<Integer, RoutingEntry> visibleTable(int nodeId) {
        Map<Integer, RoutingEntry> repaired = accepted.get(nodeId);
        if (repaired != null) {
            return repaired;
        }
        return originalTables.get(nodeId);
    }

    @Override
    public boolean isRowBroken(int routerId, int destId) {
        if (failed == null || completed.contains(routerId)) {
            return false;
        }
        Set<Integer> dests = affected.get(routerId);
        return dests != null && dests.contains(destId);
    }

    @Override
    public boolean canEditRow(int routerId, int destId) {
        return isRowBroken(routerId, destId);
    }

    @Override
    public boolean showCost() {
        return true;
    }

    @Override
    public boolean showWeights() {
        return true;
    }

    @Override
    public boolean showPacket() {
        return false;
    }

    @Override
    public NetworkLink failedLink() {
        return failed;
    }

    @Override
    public List<Integer> neighborChoices(int nodeId) {
        return live.neighborIds(nodeId);
    }

    public Set<Integer> completedNodes() {
        return Collections.unmodifiableSet(completed);
    }

    public String formatLink(NetworkLink link) {
        return original.node(link.a()).label() + " — " + original.node(link.b()).label()
                + "  (w=" + link.weight() + ")";
    }

    public String neighborHint(int nodeId) {
        StringBuilder sb = new StringBuilder();
        sb.append("Neighbors of ").append(original.node(nodeId).label()).append(":\n");
        sb.append("Bellman-Ford: new cost = link weight + neighbor's cost to dest.\n");
        sb.append("Pick the neighbor with the smallest sum.\n\n");
        for (int hop : live.neighborIds(nodeId)) {
            int weight = live.weightBetween(nodeId, hop);
            sb.append("  ").append(original.node(hop).label())
                    .append("   weight ").append(weight).append('\n');
        }
        return sb.toString();
    }

    private static Map<Integer, RoutingEntry> copyTable(Map<Integer, RoutingEntry> table) {
        return new LinkedHashMap<>(table);
    }

    private static Map<Integer, Map<Integer, RoutingEntry>> hostTables(NetworkMap map) {
        Map<Integer, Map<Integer, RoutingEntry>> tables = new LinkedHashMap<>();
        for (NetworkNode node : map.routers()) {
            Map<Integer, RoutingEntry> full = ShortestPaths.weightedTableFrom(map, node.id());
            Map<Integer, RoutingEntry> hostRows = new LinkedHashMap<>();
            for (NetworkNode host : map.hosts()) {
                hostRows.put(host.id(), full.get(host.id()));
            }
            tables.put(node.id(), hostRows);
        }
        return tables;
    }

    private static boolean hasInterestingFailure(NetworkMap map,
                                                 Map<Integer, Map<Integer, RoutingEntry>> tables) {
        for (NetworkLink link : map.links()) {
            if (!map.withoutLink(link).isConnected()) {
                continue;
            }
            for (NetworkNode router : map.routers()) {
                for (NetworkNode host : map.hosts()) {
                    if (walkUsesLink(map, tables, router.id(), host.id(), link)) {
                        return true;
                    }
                }
            }
        }
        return false;
    }

    private static boolean walkUsesLink(NetworkMap map,
                                        Map<Integer, Map<Integer, RoutingEntry>> tables,
                                        int from, int dest, NetworkLink link) {
        int cur = from;
        int guard = map.nodes().size() + 2;
        while (cur != dest && guard-- > 0) {
            Map<Integer, RoutingEntry> table = tables.get(cur);
            if (table == null) {
                return false;
            }
            RoutingEntry entry = table.get(dest);
            if (entry == null) {
                return false;
            }
            int next = entry.nextHop();
            if (link.connects(cur) && link.connects(next)) {
                return true;
            }
            cur = next;
        }
        return false;
    }
}

File: web/src/controller/BellmanFordSession.js
import { JavaRandom } from "../rng.js";
import { NetworkMap } from "../model/NetworkMap.js";
import { RoutingEntry } from "../model/RoutingEntry.js";
import { ShortestPaths } from "../model/ShortestPaths.js";
import { CheckResult, SelectResult } from "./GameSession.js";

export const Phase = {
  PICK_LINK: "PICK_LINK",
  REPAIR: "REPAIR",
  DONE: "DONE",
};

export const FailResult = {
  OK: "OK",
  ALREADY_FAILED: "ALREADY_FAILED",
  DISCONNECTS: "DISCONNECTS",
  NOT_USED: "NOT_USED",
};

export class BellmanFordSession {
  constructor(nodeCount, seed) {
    const random = new JavaRandom(seed);
    let chosen = null;
    let tables = null;
    for (let attempt = 0; attempt < 20; attempt++) {
      const candidate = NetworkMap.random(nodeCount, random);
      const built = hostTables(candidate);
      if (hasInterestingFailure(candidate, built)) {
        chosen = candidate;
        tables = built;
        break;
      }
      if (chosen == null) {
        chosen = candidate;
        tables = built;
      }
    }
    this._original = chosen;
    this._live = chosen;
    this._failed = null;
    this._originalTables = tables;
    this._accepted = new Map();
    this._affected = new Map();
    this._completed = new Set();
    this._selectedNode = null;
    this._phase = Phase.PICK_LINK;
  }

  phase() {
    return this._phase;
  }

  remainingRepairs() {
    let n = 0;
    for (const id of this._affected.keys()) {
      if (!this._completed.has(id)) {
        n++;
      }
    }
    return n;
  }

  affectedRouterCount() {
    return this._affected.size;
  }

  failLink(link) {
    if (this._failed != null) {
      return FailResult.ALREADY_FAILED;
    }
    const canonical = this._original.linkBetween(link.a(), link.b());
    if (canonical == null) {
      return FailResult.NOT_USED;
    }
    const next = this._original.withoutLink(canonical);
    if (!next.isConnected()) {
      return FailResult.DISCONNECTS;
    }
    const broken = new Map();
    for (const router of this._original.routers()) {
      const dests = [];
      for (const host of this._original.hosts()) {
        if (walkUsesLink(this._original, this._originalTables, router.id(), host.id(), canonical)) {
          dests.push(host.id());
        }
      }
      if (dests.length > 0) {
        broken.set(router.id(), new Set(dests));
      }
    }
    if (broken.size === 0) {
      return FailResult.NOT_USED;
    }
    this._failed = canonical;
    this._live = next;
    this._affected = broken;
    this._phase = Phase.REPAIR;
    for (const router of this._original.routers()) {
      if (!this._affected.has(router.id())) {
        this._completed.add(router.id());
        this._accepted.set(router.id(), copyTable(this._originalTables.get(router.id())));
      }
    }
    return FailResult.OK;
  }

  map() {
    return this._original;
  }

  selectedNode() {
    return this._selectedNode;
  }

  selectNode(nodeId) {
    if (this._original.node(nodeId).isHost()) {
      return SelectResult.HOST;
    }
    this._selectedNode = nodeId;
    return SelectResult.OK;
  }

  check(nodeId, playerTable) {
    if (this._failed == null) {
      return CheckResult.INCOMPLETE;
    }
    if (this._original.node(nodeId).isHost()) {
      return CheckResult.INCOMPLETE;
    }
    if (this._completed.has(nodeId)) {
      return CheckResult.CORRECT;
    }
    const dests = this._affected.get(nodeId);
    if (dests == null || dests.size === 0) {
      return CheckResult.CORRECT;
    }
    for (const dest of dests) {
      if (playerTable.get(dest) == null) {
        return CheckResult.INCOMPLETE;
      }
    }
    for (const dest of dests) {
      const player = playerTable.get(dest);
      if (!ShortestPaths.isLegalWeightedNextHop(this._live, nodeId, player.nextHop(), dest)) {
        return CheckResult.WRONG_HOP;
      }
    }
    const table = copyTable(this._originalTables.get(nodeId));
    for (const dest of dests) {
      const hop = playerTable.get(dest).nextHop();
      const cost = ShortestPaths.weightedDistance(this._live, nodeId, dest);
      table.set(dest, new RoutingEntry(dest, hop, cost));
    }
    this._accepted.set(nodeId, table);
    this._completed.add(nodeId);
    if (this.remainingRepairs() === 0) {
      this._phase = Phase.DONE;
    }
    return CheckResult.CORRECT;
  }

  isCompleted(nodeId) {
    return this._completed.has(nodeId);
  }

  isUnlocked(nodeId) {
    return this._original.node(nodeId).isRouter();
  }

  toldNextHop(routerId) {
    const table = this.visibleTable(routerId);
    if (table == null) {
      return null;
    }
    const entry = table.get(this._original.destId());
    return entry == null ? null : entry.nextHop();
  }

  packetArrived() {
    return false;
  }

  playerPathSuboptimal() {
    return false;
  }

  playerPath() {
    return [];
  }

  optimalPath() {
    return [];
  }

  packetNode() {
    return this._original.sourceId();
  }

  acceptedTable(nodeId) {
    return this._accepted.get(nodeId);
  }

  visibleTable(nodeId) {
    const repaired = this._accepted.get(nodeId);
    if (repaired != null) {
      return repaired;
    }
    return this._originalTables.get(nodeId);
  }

  isRowBroken(routerId, destId) {
    if (this._failed == null || this._completed.has(routerId)) {
      return false;
    }
    const dests = this._affected.get(routerId);
    return dests != null && dests.has(destId);
  }

  canEditRow(routerId, destId) {
    return this.isRowBroken(routerId, destId);
  }

  showCost() {
    return true;
  }

  showWeights() {
    return true;
  }

  showPacket() {
    return false;
  }

  failedLink() {
    return this._failed;
  }

  neighborChoices(nodeId) {
    return this._live.neighborIds(nodeId);
  }

  completedNodes() {
    return this._completed;
  }

  formatLink(link) {
    return `${this._original.node(link.a()).label()} — ${this._original.node(link.b()).label()}  (w=${link.weight()})`;
  }

  neighborHint(nodeId) {
    const lines = [];
    lines.push(`Neighbors of ${this._original.node(nodeId).label()}:`);
    lines.push("Bellman-Ford: new cost = link weight + neighbor's cost to dest.");
    lines.push("Pick the neighbor with the smallest sum.");
    lines.push("");
    for (const hop of this._live.neighborIds(nodeId)) {
      const weight = this._live.weightBetween(nodeId, hop);
      lines.push(`  ${this._original.node(hop).label()}   weight ${weight}`);
    }
    return lines.join("\n");
  }
}

function copyTable(table) {
  return new Map(table);
}

function hostTables(map) {
  const tables = new Map();
  for (const node of map.routers()) {
    const full = ShortestPaths.weightedTableFrom(map, node.id());
    const hostRows = new Map();
    for (const host of map.hosts()) {
      hostRows.set(host.id(), full.get(host.id()));
    }
    tables.set(node.id(), hostRows);
  }
  return tables;
}

function hasInterestingFailure(map, tables) {
  for (const link of map.links()) {
    if (!map.withoutLink(link).isConnected()) {
      continue;
    }
    for (const router of map.routers()) {
      for (const host of map.hosts()) {
        if (walkUsesLink(map, tables, router.id(), host.id(), link)) {
          return true;
        }
      }
    }
  }
  return false;
}

function walkUsesLink(map, tables, from, dest, link) {
  let cur = from;
  let guard = map.nodes().length + 2;
  while (cur !== dest && guard-- > 0) {
    const table = tables.get(cur);
    if (table == null) {
      return false;
    }
    const entry = table.get(dest);
    if (entry == null) {
      return false;
    }
    const next = entry.nextHop();
    if (link.connects(cur) && link.connects(next)) {
      return true;
    }
    cur = next;
  }
  return false;
}
